documentation/src/docs/controller-service.md

```diff
@@ -154,296 +151,276 @@ Controller Process Diagram
+
+#### Sealed Segments Maps
+Once the Stream Segments are sealed, Controller needs to store additional information about the Stream Segment. Presently, we have two types of information:
 
-#### Sealed Segments Record
- Since the segment table is append only, any additional information
- that we need to persist when a segment is sealed is stored in sealed
- segments record. Presently, it simple contains a map of segment number
- to its sealed size.
+ - Epoch, the Stream Segment was sealed in.
+ - Size of the Stream Segment at the time of sealing.
+
+ These records have two different characteristics and are used in different types of queries. For example;
+
+ 1. Sealing epoch is important for querying successor Stream Segments. For each Stream Segment we store its sealing epoch directly in the metadata store.
+ 2. Stream Segment sizes are used during truncation workflows. For sealed sizes, we store it in a map of Segment to size at the time of sealing.
+ 3. Successor queries are performed on a single Stream Segment whereas truncation workflows work on a group of Stream Segments.
+
+ This ensures that during truncation we are able to retrieve sealed sizes for multiple Stream Segments with minimal number of calls into underlying metadata store. Since we could have arbitrarily large number of Stream Segments that have been sealed away, we cannot store all of the information in a single map and hence we shard the map and store it. The sharding function we use is to hash the creation epoch and get the shard number.
 
 The following are the Transaction Related metadata records:
 
-   - **Active Transactions**: Each new Transaction is created under this znode. This stores metadata
- corresponding to each Transaction as *ActiveTransactionRecord*. Once a
+   - **Active Transactions**: Each new Transaction is created under the znode. This stores metadata
+ corresponding to each Transaction as *Active Transaction Record*. Once a
  Transaction is completed, a new node is created under the global
- Completed Transaction znode and removed from under the stream specific
- _active_ Transaction node.
+ _Completed Transaction_ znode and removed from under the Stream specific
+ _Active Transaction_ node.
 
-   - **Completed Transactions**: All completed transactions for all streams are moved under this single
+   - **Completed Transactions**: All completed transactions for all Streams are moved under this single
  znode upon completion (via either commit or abort paths). We can
  subsequently garbage collect these values periodically following any
  collection scheme we deem fit. We have not implemented any scheme at
  this point though.
 
 ### Stream Store Caching
 
-Since there could be multiple concurrent requests for a given stream
+#### In-memory Cache
+Since there could be multiple concurrent requests for a given Stream
 being processed by same Controller instance, it is suboptimal to read
 the value by querying Zookeeper every time. So we have introduced an
```

Comment:
@shiveshr true, but we need to argue why that's not a (consistency) problem as the value can in principle change under the hood in zk.

Reply:
@shiveshr any updates on this?

Reply:
Please add following lines:
```The cache can contain both mutable and immutable values. Immutable values, by definition are not a problem. For mutable values, we have introduced a notion of OperationContext and for each new operation, which ensures that during an operation we lazily load latest value of entities into the cache and then use them for all computations within that Operation's context. Please refer to OperationContext (TODO: @Aparnarr add link) for more details. ```

Reply:
@shiveshr edited please check